sdk/core/azure-core/src/http/response.cpp

```diff
@@ -1,35 +1,57 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+#include <cctype>
+#include <http/http.hpp>
 #include <map>
 #include <string>
 #include <vector>
 
-#include <http/http.hpp>
-
 using namespace Azure::Core::Http;
 
-uint16_t Response::getStatusCode()
-{
-  return m_statusCode;
-}
+HttpStatusCode Response::GetStatusCode() { return m_statusCode; }
 
-std::string const& Response::getReasonPhrase()
-{
-  return m_reasonPhrase;
-}
+std::string const& Response::GetReasonPhrase() { return m_reasonPhrase; }
+
+std::map<std::string, std::string> const& Response::GetHeaders() { return this->m_headers; }
 
-std::map<std::string, std::string> const& Response::getHeaders()
+std::vector<uint8_t>& Response::GetBodyBuffer() { return m_bodyBuffer; }
+
+namespace {
+inline bool IsStringEqualsIgnoreCase(std::string const& a, std::string const& b)
 {
-  return this->m_headers;
+  auto alen = a.length();
+  auto blen = b.length();
+
+  if (alen != blen)
+  {
+    return false;
+  }
+
+  for (size_t index = 0; index < alen; index++)
+  {
+    // TODO: tolower is bad for some charsets, see if this can be enhanced
+    if (std::tolower(a.at(index)) != std::tolower(b.at(index)))
+    {
+      return false;
+    }
+  }
+  return true;
 }
+} // namespace
 
-BodyStream* Response::getBodyStream()
+void Response::AddHeader(std::string const& name, std::string const& value)
 {
-  return m_bodyStream;
+  if (IsStringEqualsIgnoreCase("Content-Length", name))
+  {
+    // whenever this header is found, we reserve the value of it to be pre-allocated to write
+    // response
+    m_bodyBuffer.reserve(std::stol(value));
+  }
+  this->m_headers.insert(std::pair<std::string, std::string>(name, value));
```

Comment:
```suggestion
  this->m_headers.emplace(name, value);
```

We may want to conditionally call `insert_or_assign` if it is available but that's a C++17 feature.